File: bert/end2end/src/mrpc.cpp
#include <iostream>
#include <vector>
#include <ctime>
#include <unordered_map>
#include <string>
#include <fstream>
#include <migraphx/literal.hpp>
#include <migraphx/operators.hpp>
#include <migraphx/program.hpp>
#include <migraphx/instruction.hpp>
#include <migraphx/quantization.hpp>
#include <migraphx/onnx.hpp>
#include <migraphx/generate.hpp>
#include <migraphx/cpu/target.hpp>
#include <migraphx/gpu/target.hpp>
#include <migraphx/gpu/hip.hpp>

using namespace std::chrono;

migraphx::program load_onnx_file(std::string file_name) {
    auto prog = migraphx::parse_onnx(file_name);
    std::cout << "prog = " << std::endl;
    std::cout << prog << std::endl;
    return prog;
}

template <class T>
auto get_hash(const T& x)
{
    return std::hash<T>{}(x);
}

void parse_sentence(const std::string& sent, std::vector<int64_t>& vec_feature)
{
    size_t pos = 0, pos_next;
    size_t index = 0;
    while ((pos_next = sent.find(',', pos)) != std::string::npos)
    {
        auto word_feature = sent.substr(pos, pos_next);
        vec_feature.push_back(std::stoll(word_feature));
        pos = pos_next + 1;
    }
    vec_feature.push_back(std::stoll(sent.substr(pos)));
    vec_feature.push_back(102);
}

int parse_line(std::string& line, std::size_t sent_size, 
        std::unordered_map<std::string, std::vector<int64_t>>& input_map)
{
    auto& vec_feature = input_map["input.1"];
    auto& vec_id = input_map["input.3"];
    auto& seg_id = input_map["2"];
    vec_feature.clear();
    vec_id.clear();
    seg_id.clear();

    size_t pos = line.find('\t');
    int label = std::stoi(line.substr(0, pos));

    ++pos;
    size_t pos_next = line.find('\t', pos);
    vec_feature.push_back(101);
    parse_sentence(line.substr(pos, pos_next), vec_feature);
    vec_id.resize(vec_feature.size(), 0);

    pos = pos_next + 1;
    parse_sentence(line.substr(pos), vec_feature);
    vec_id.resize(vec_feature.size(), 1);
    seg_id.resize(vec_feature.size(), 1);

    vec_feature.resize(sent_size, 0);
    vec_id.resize(sent_size, 0);
    seg_id.resize(sent_size, 0);

    return label;
}

void print_vec(std::vector<int64_t>& vec)
{
    for (auto v : vec)
    {
        std::cout << v << ", ";
    }
    std::cout << std::endl;
}

int main(int argc, char **argv) {
    if (argc != 4) {
        std::cout << "Usage: " << argv[0] << " onnx_file input_data cpu/gpu" << std::endl;
        return 0;
    }

    std::string target_name("cpu");
    if (std::string(argv[3]) == std::string("gpu"))
    {
        target_name = "gpu";
    }

    auto prog = load_onnx_file(argv[1]);
    //migraphx::quantize(prog);
    //std::cout << "quantized prog = " << std::endl;
    std::cout << prog << std::endl;
    if (target_name == "cpu")
    {
        prog.compile(migraphx::cpu::target{});
    }
    else
    {
        prog.compile(migraphx::gpu::target{});
    }

    std::size_t batch_size = 1;
    auto param_shapes = prog.get_parameter_shapes();
    if (param_shapes.count("input.1") > 0)
    {
        batch_size = prog.get_parameter_shapes()["input.1"].lens()[0];
    }

    std::ifstream ifs(argv[2]);
    if (!ifs.is_open())
    {
        std::cout << "Open file " << argv[2] << " error!" << std::endl;
        return 1;
    }

    std::string line;
    std::getline(ifs, line);
    migraphx::program::parameter_map m;
    std::size_t accu_count = 0, total_count = 0;
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    while (true)
    {
        std::unordered_map<std::string, std::vector<int64_t>> input_map;
        input_map["input.1"];
        input_map["input.3"];
        input_map["2"];
        std::unordered_map<std::string, std::vector<int64_t>> sent_tokens;
        sent_tokens["input.1"];
        sent_tokens["input.3"];
        sent_tokens["2"];
        std::vector<int> vec_labels;

        for (std::size_t batch_no = 0; batch_no < batch_size; batch_no++)
        {
            std::getline(ifs, line);
            if (line.empty())
            {
                break;
            }
            int label = parse_line(line, 128, sent_tokens);
            vec_labels.push_back(label);
            input_map["input.1"].insert(input_map["input.1"].end(), sent_tokens["input.1"].begin(),
                    sent_tokens["input.1"].end());
            input_map["input.3"].insert(input_map["input.3"].end(), sent_tokens["input.3"].begin(),
                    sent_tokens["input.3"].end());
            input_map["2"].insert(input_map["2"].end(), sent_tokens["2"].begin(),
                    sent_tokens["2"].end());
        }
       
        if (line.empty())
        {
            break;
        }

        for (auto &&x : prog.get_parameter_shapes())
        {
            //std::cout << "gpu input: " << x.first << ", shape = " << x.second << std::endl;
            migraphx::argument argu{};
            if (input_map.count(x.first) > 0)
            {
                argu = migraphx::argument(x.second, input_map[x.first].data());
            }
            else
            {
                argu = migraphx::generate_argument(x.second, get_hash(x.first));
            }

            if (target_name == "cpu")
            {
                m[x.first] = argu;
            }
            else
            {
                m[x.first] = migraphx::gpu::to_gpu(argu);
            }
        }

        migraphx::argument result{};
        if (target_name == "cpu")
        {
            result = prog.eval(m);
        }
        else
        {
            result = migraphx::gpu::from_gpu(prog.eval(m));
        }
        std::vector<float> vec_output;
        result.visit([&](auto output) { vec_output.assign(output.begin(), output.end()); });

        for (std::size_t batch_no = 0; batch_no < batch_size; ++batch_no)
        {
            std::cout << "[" << vec_output[2 * batch_no] << ", " << vec_output[2 * batch_no + 1] << "]" << std::endl;
            int calc_label = (vec_output[2 * batch_no] >= vec_output[2 * batch_no + 1]) ? 0 : 1;
            accu_count += (calc_label == vec_labels[batch_no]) ? 1 : 0;
            ++total_count;
        }
    }

    std::cout << "accuracy rate = " << 1.0 * accu_count / total_count << std::endl;

    high_resolution_clock::time_point t2 = high_resolution_clock::now();
    
    milliseconds ms = duration_cast<milliseconds>(t2 - t1);
    std::cout << "It takes " << ms.count() << " ms" << std::endl;
    return 0;
}

File: bert/end2end/python/mrpc.py
from io import open
import string
import re
import random
import sys
import migraphx
import time
import numpy as np

def parse_line(line):
    tokens = line.strip().split('\t')
    label = int(tokens[0])
    token_ids = list(map(int, tokens[1].strip().split(',')))
    token_ids.insert(0, 101)
    token_ids.append(102)
    sent_id = [0] * len(token_ids)
    sent2_token_ids = list(map(int, tokens[2].strip().split(',')))
    sent2_token_ids.append(102)
    sent2_id = [1] * len(sent2_token_ids)
    token_ids.extend(sent2_token_ids)
    sent_id.extend(sent2_id)
    seg_id = [1] * len(sent_id)

    if len(sent_id) < 128:
        patch = [0] * (128 - len(sent_id))
        token_ids.extend(patch)
        sent_id.extend(patch)
        seg_id.extend(patch)

    return (label, token_ids, sent_id, seg_id)

def check_lines(model, scratch_output, target_name, batch_lines):
    input_map = {"input.1": [],
                 "input.3": [],
                 "2": []}
    batch_labels = []
    for l in batch_lines:
        label, token_ids, sent_id, seg_id = parse_line(l)
        input_map["input.1"].extend(token_ids)
        input_map["input.3"].extend(sent_id)
        input_map["2"].extend(seg_id)
        batch_labels.append(label)
 
    model_param_map = {}
    args = []
    for key, value in model.get_parameter_shapes().items():
        if key in scratch_output:
            model_param_map[key] = scratch_output[key]
        else:
            if key in input_map:
                args.append(np.array(input_map[key], dtype = np.longlong).reshape(value.lens()))
            else:
                hash_val = hash(key) % (2 ** 32 - 1)
                np.random.seed(hash_val)
                args.append(np.random.randn(value.elements()).astype(np.single).reshape(value.lens()))
            model_param_map[key] = migraphx.argument(args[-1])

    result = np.array(model.run(model_param_map))

    print("{}".format(result))
    output = np.squeeze(np.reshape(result, (1, -1)), axis=0).tolist()

    correct_num = 0;
    for i in range(len(batch_lines)):
        calc_label = 0 if output[2 * i] >= output[2 * i + 1] else 1
        res = 1 if calc_label == batch_labels[i] else 0
        correct_num += res

    return correct_num

def main():
    if len(sys.argv) != 4:
        print("Usage: python mrpc.py file.onnx input.tsv cpu/gpu")
        exit(1)

    # load model
    model = migraphx.parse_onnx(sys.argv[1])

    # read input file
    lines = open(sys.argv[2]).read().strip().split('\n')

    # get target name
    target_name = "cpu"
    if sys.argv[3] == "gpu":
        target_name = "gpu"

    model.compile(migraphx.get_target(target_name))
    batch_size = model.get_parameter_shapes()["input.1"].lens()[0]

    # wrap up output and scratch memory to be reused
    scratch_output = {}
    for key, value in model.get_parameter_shapes().items():
        if key == "scratch" or key == "output":
            hash_val = hash(key) % (2 ** 32 - 1)
            np.random.seed(hash_val)
            scratch_output[key] = migraphx.argument(np.random.randn(value.elements()).astype(np.single).reshape(value.lens()))

    # first line is useless information
    accu_num = 0;
    indices = list(range(len(lines) - 1))
    indices = indices[1:(len(lines) - 1) : batch_size]
    start = time.time()
    for i in indices:
        chk_res = check_lines(model, scratch_output, target_name, lines[i : i + batch_size])
        accu_num += chk_res
    end = time.time()
    
    accu_rate = 1.0 * accu_num / (len(lines) - 1)
    print("accuracy rate = {}".format(accu_rate))
    print("elapsed time = {}".format(end - start))

if __name__ == "__main__":
    main()
